src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentComplexAttributeContentPass.cs

```diff
@@ -36,88 +36,86 @@ private void ProcessAttributes(TagHelperIntermediateNode node)
-            cSharpNode.Children.Count > 1)
+        else if (node.Children is [CSharpExpressionIntermediateNode { Children.Count: > 1 } cSharpNode])
         {
             // This case can be hit when the attribute has an explicit @ inside, which
             // 'escapes' any special sugar we provide for codegen.
             //
             // There's a special case here for explicit expressions. See https://github.com/aspnet/Razor/issues/2203
             // handling this case as a tactical matter since it's important for lambdas.
-            if (cSharpNode.Children.Count == 3 &&
-                cSharpNode.Children[0] is IntermediateToken token0 &&
-                cSharpNode.Children[2] is IntermediateToken token2 &&
-                token0.Content == "(" &&
-                token2.Content == ")")
+            if (cSharpNode.Children is [IntermediateToken { Content: "(" }, _, IntermediateToken { Content: ")" }])
             {
                 cSharpNode.Children.RemoveAt(2);
                 cSharpNode.Children.RemoveAt(0);
-
-                // We were able to simplify it, all good.
-                return true;
             }
-
-            return false;
+            else
+            {
+                removeNode = true;
+                issueDiagnostic = true;
+            }
         }
-        else if (node.Children.Count == 1 &&
-            node.Children[0] is CSharpCodeIntermediateNode)
+        else if (node.Children is [CSharpCodeIntermediateNode])
         {
             // This is the case when an attribute contains a code block @{ ... }
             // We don't support this.
-            return false;
+            removeNode = true;
+            issueDiagnostic = true;
+        }
+        else if (node.Children is [CSharpExpressionIntermediateNode, HtmlContentIntermediateNode { Children: [IntermediateToken { Content: "." }] }])
+        {
+            // This is the case when an attribute contains something like "@MyEnum."
+            // We simplify this to remove the "." so that tooling can provide completion on "MyEnum"
+            // in case the user is in the middle of typing
+            node.Children.RemoveAt(1);
+
+            // We still want to issue a diagnostic, even though we simplified, because ultimately
+            // we don't support this, so if the user isn't typing, we can't let this through
+            issueDiagnostic = true;
         }
         else if (node.Children.Count > 1)
         {
             // This is the common case for 'mixed' content
-            return false;
+            removeNode = true;
+            issueDiagnostic = true;
         }
 
-        return true;
+        if (issueDiagnostic)
+        {
+            parent.Diagnostics.Add(ComponentDiagnosticFactory.Create_UnsupportedComplexContent(
+                node,
+                attributeName));
+        }
+        if (removeNode)
+        {
+            parent.Children.Remove(node);
```

Comment:
It would be better for tooling if this just removed the attribute value, not the whole attribute node, but the intermediate tree doesn't really have a structure like that, and removing all child nodes felt like it would probably just end up breaking something further in the pipeline. I logged https://github.com/dotnet/razor/issues/9346 to hopefully improve this in future.